### DISCUSSION TITLE
Getting at notdef unicode codepoints

### DISCUSSION THREAD

**Discussion by talltyler** [harfbuzz/harfbuzzjs#71] [Q&A] [answered]
From my understanding `var result = buffer.json(font);`
should return an array of https://harfbuzz.github.io/harfbuzz-hb-buffer.html#hb_glyph_info_t
but the current implementation is only returning clusters without the original unicode codepoints

There is other unused data inside of `infos` I thought it might be in
```
var infosPtr = exports.hb_buffer_get_glyph_infos(ptr, 0);
var infosPtr32 = infosPtr / 4;
var infos = heapu32.subarray(infosPtr32, infosPtr32 + 5 * length);
```
but none of this additional unreturned data seems to be what I'm looking for.

I'm looking to use the unicode codepoints from the original buffer to handle fallbacks for glyphs that are not available in the font. 
Currently the glyphId is returned as 0 for all of these cases which isn't very helpful.

I'm open to alternative ways of handling this if you all have better ideas for ways to connect the original text with the shaped output.

**Comment by chearon** [comment]
In the `result` array each object has `cl` for the UTF-16 offset into the string and `g` for the glyph ID in the font. If the glyph ID is 0 you need to do fallback on the unicode cluster it corresponds to. Maybe you were expecting `cl` to be a full code point value rather than an offset? If you post a full example I could be of better help.

**Comment by talltyler** [comment]
For now I've just been modifying the provided examples and trying to get strings with emojis to return something I can render. From my understanding cl is the cluster index which for single byte characters will work fine but characters like 👱🏽‍♂️ all of this breaks down because the cluster indexes become disconnected from the original buffer strings with the multibyte characters. I'd tried a few things to get your suggested approach to work by counting the byte length of different characters and modifying the cl values into the original strings index but there are a lot of edge cases. This isn't feeling like the correct path to take and I just want to check if maybe there is another way to access this data.

**Comment by chearon** [comment]
If I drag Noto Color Emoji into the [harfbuzzjs demo](https://harfbuzz.github.io/harfbuzzjs/) and type "a👱🏽‍♂️b" into the input, it gives me this array:

```
[
  {"g":0,"cl":0,"ax":1245,"ay":0,"dx":0,"dy":0,"flags":0},
  {"g":2539,"cl":1,"ax":1245,"ay":0,"dx":0,"dy":0,"flags":0},
  {"g":0,"cl":8,"ax":1245,"ay":0,"dx":0,"dy":0,"flags":0}
]
```

Since `'👱🏽‍♂️'.length === 7`, all of those indices are correct. What are the disconnected indices you're seeing?

**Comment by talltyler** [comment]
I added code based on my understanding of your logic to print out the string representation of cluster along with the other relevant information which works for the case you are talking about. Using 'Noto Color Emoji' I get this result using the same input `a👱🏽‍♂️b`. 
```[
  {"g": 0, "cl": 0, "ax": 1245, "flags": 0, "string": "a"},
  {"g": 725, "cl": 1, "ax": 1245, "flags": 0, "string": "👱🏽‍♂️"},
  {"g": 0, "cl": 8, "ax": 1245, "flags": 0, "string": "b"}
]
```
What I've been looking to do it the opposite. Using fonts like Roboto-Black.ttf that is included in the examples I would like to be able to render both what is supported by Roboto and other multibyte characters like emojis. With this Roboto I get this output from the same string and code from above.
```[
  {"g": 68, "cl": 0, "ax": 530, "flags": 0, "string": "a"},
  {"g": 0, "cl": 1, "ax": 365, "flags": 0, "string": "👱"},
  {"g": 0, "cl": 3, "ax": 365, "flags": 1, "string": "🏽"},
  {"g": 3, "cl": 5, "ax": 0, "flags": 1, "string": "‍"},
  {"g": 0, "cl": 6, "ax": 365, "flags": 1, "string": "♂"},
  {"g": 3, "cl": 7, "ax": 0, "flags": 1, "string": "️"},
  {"g": 69, "cl": 8, "ax": 562, "flags": 0, "string": "b"}
]
```
I was assuming this output was expected and I needed to find a way to take the glyphs unsupported by this font out of this result but maybe what you are telling me is that this shaped result is wrong and I should always be seeing 3 clusters but 7?

I've been trying to work with these results as is and this is the basis for my questions. I was thinking I could maybe ignore cl 5 and 7, then anything with the same ax value could maybe be part of the same multibyte character but there are layers of assumptions here and it gets worse with longer strings and multiple emojis which is why I wanted to link the original input string back to these results more directly, there is still a connection here but the results are hard for me to reason with.

I'm trying to render text in a variety of fonts that don't always have all the glyphs they need but have fallbacks to other fonts so that the original input is always rendered. I was trying todo this one cluster at a time based on the shaped results but maybe I should be thinking about this differently.

**Reply by behdad** [reply]
> I've been trying to work with these results as is and this is the basis for my questions. I was thinking I could maybe ignore cl 5 and 7, then anything with the same ax value could maybe be part of the same multibyte character but there are layers of assumptions here and it gets worse with longer strings and multiple emojis which is why I wanted to link the original input string back to these results more directly, there is still a connection here but the results are hard for me to reason with.
> 
> I'm trying to render text in a variety of fonts that don't always have all the glyphs they need but have fallbacks to other fonts so that the original input is always rendered. I was trying todo this one cluster at a time based on the shaped results but maybe I should be thinking about this differently.

Indeed. You should be using what's called a Unicode grapheme cluster as a unit of font fallback instead.  If you lack library support for that though, you can *try* using the HarfBuzz clusters, but that will come with limitations like what you're observing...

That said, for the emoji sequences I expect that we always cluster the full example into one cluster. I need to study this more to see why it's not happening here.

**Reply by behdad** [reply]
Shaping with hb-shape and Roboto-Regular.ttf, I get the same cluster number for all glyphs for the 👱🏽‍♂️ sequence. I expect you to get the same.

**Reply by behdad** [reply]
Can you try this with the `hb-shape` command-line tool?

**Reply by behdad** [reply]
```
$ build/util/hb-shape Roboto-Regular.ttf 👱🏽‍♂️
[gid0=0+908|gid0=0+908|gid4=0+0|gid0=0+908|gid4=0+0]
```

**Reply by talltyler** [reply]
Using Roboto-Black.ttf from the examples folder I get
[.notdef=0+748|.notdef=0+748|space=0+0|.notdef=0+748|space=0+0]

This is all from a fresh github download of harfbuzz a day before my first message but I had only compiled the WASM file originally. This is the first time I've compiled the full source version. If there is a version I should be working from I'm happy to try something else. All of the builds I've done so far are fallowing the documentation in the repo.

**Reply by behdad** [reply]
Okay so what you get from hb-shape is good. Now to debug what config in harfbuzzjs is causing the subpar result there.

**Comment by talltyler** [comment, marked as answer]
So that you know I've also tried setClusterLevel of 0, 1 and 2 which from my understanding is the main thing that controls which grapheme cluster boundary is used.
The output for one and two for the same as I posted before but setting the value to 0 give this response which seems worse but the docs also recommend not using this cluster level.
```
[
  { "g": 68, "cl": 0, "ax": 530, "flags": 0, "string": "a" },
  { "g": 0, "cl": 1, "ax": 365, "flags": 0, "string": "\ud83d" },
  { "g": 0, "cl": 1, "ax": 365, "flags": 0, "string": "\ud83d" },
  { "g": 3, "cl": 1, "ax": 0, "flags": 0, "string": "\ud83d" },
  { "g": 0, "cl": 1, "ax": 365, "flags": 0, "string": "\ud83d" },
  { "g": 3, "cl": 1, "ax": 0, "flags": 0, "string": "👱🏽‍♂️" },
  { "g": 69, "cl": 8, "ax": 562, "flags": 0, "string": "b" }
]
```
If you have suggested of other API methods to look into please let me know.

**Reply by behdad** [reply]
> setting the value to 0 give this response which seems worse but the docs also recommend not using this cluster level.

This is what you want, and the output looks expected. Where do the docs recommend not using it?

**Reply by talltyler** [reply]
"Therefore, level 1 clustering is recommended for code that is not required to implement backward compatibility with the old HarfBuzz." https://harfbuzz.github.io/working-with-harfbuzz-clusters.html

If I switch the cluster level to 0, you are saying I should only render the last items from every cluster group or are you reading something different into these results.

**Reply by behdad** [reply]
> "Therefore, level 1 clustering is recommended for code that is not required to implement backward compatibility with the old HarfBuzz." https://harfbuzz.github.io/working-with-harfbuzz-clusters.html

Thanks. I should improve that.

> If I switch the cluster level to 0, you are saying I should only render the last items from every cluster group or are you reading something different into these results.

No. What you need to do, since you want to implement font-fallback is to check the glyphs for each cluster, and if you found glyph-0 (notdef) in a cluster, then perform fallback for the entire cluster.

**Reply by talltyler** [reply]
Thank you, I think we can resolve this conversation. I'm pretty sure I can get this working based on this knowledge, thanks again for taking the time to explain all of this.
